### DISCUSSION TITLE
How to generate federation v2 compatible schemas via gradle in 6.x ?

### DISCUSSION THREAD

**Discussion by roookeee** [ExpediaGroup/graphql-kotlin#1644] [General] [answered]
I am getting the following error when using federation v2 annotations during my gradle build:

```
Execution failed for task ':graphqlGenerateSDL'.
> There was a failure while executing work items
   > A failure occurred while executing com.expediagroup.graphql.plugin.gradle.actions.GenerateSDLAction
      > Attempting to use Federation v2 shareable directive while generating Federation v1 compatible schema
```

I had no luck finding a property to set the opt in flag for the gradle plugin action. How do I allow the Gradle action to generate a federation v2 schema? I checked all actions and think its an oversight as its settable e.g. in the Spring server in 6.x.

I know the default will change in 7.x, but that version is still in alpha. Having the ability to override the opt in for the gradle task would be an improvement regardless.

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 

Federation v2 is supported with an opt-in mechanism in `graphql-kotlin` v6 (note: we will be releasing v6.3.5 that contains a fed bug fix soon) and it defaults to v1 if flag is not specified. 

Plugins rely on a service loader mechanism to generate SDL based on the provided hooks (defaults to non-federated schema). Service loader mechanism is pretty flexible as it allows us to dynamically load hooks at runtime based on the available classpath. On the other hand it is also non-configurable as we cannot pass arguments in some generic fashion (as we don't know what will be loaded) - that's why don't have a plugin configuration option to distinguish between v1 and v2 and we use the same default behavior as the schema generator.

Since Fed v2 is an opt-in behavior (in `graphql-kotlin` v6, is a default in v7), provided `federated-hooks-provider` also default to use v1. If you need to generate Fed v2 SDL, then you need to provide a custom hooks provider - see [our docs](https://opensource.expediagroup.com/graphql-kotlin/docs/plugins/hooks-provider) for details. TLDR you need to create a copy of [FederatedHooksProvider](https://github.com/ExpediaGroup/graphql-kotlin/blob/6.x.x/plugins/schema/graphql-kotlin-federated-hooks-provider/src/main/kotlin/com/expediagroup/graphql/plugin/schema/hooks/FederatedSchemaGeneratorHooksProvider.kt) that sets the opt in flag to true (+ don't forget to add service loader info under `META-INF/services` as well) 

Thanks,
Derek

**Reply by samuelAndalon** [reply]
https://github.com/ExpediaGroup/graphql-kotlin/releases/tag/6.3.5
